src/cdk/overlay/overlay-ref.ts

```diff
@@ -22,12 +23,16 @@ export class OverlayRef implements PortalHost {
   private _backdropClick: Subject<any> = new Subject();
   private _attachments = new Subject<void>();
   private _detachments = new Subject<void>();
+  private _disposed = new Subject<void>();
+
+  _dispatchedKeyboardEvents = new Subject<KeyboardEvent>();
```

Comment:
Instead of having the `OverlayRef` add and remove itself from the dispatcher, a cleaner approach would be to only have it expose the stream of keyboard events. Then the `OverlayRef.keyboardEvents`  can be simplified to:

```ts
// note: don't get too caught up in the naming
return this.keyboardDispatcher.events.filter(event => event.target === this.overlayElement);
```

This less room for mistakes when managing the overlays and it also removes the circular dependency between the `OverlayRef` and the dispatcher.

Reply:
Hm I don't entirely follow.

1. I kind of see how I could remove the `Subject`; maybe if I created a new event type that paired overlays to keydowns, and then the overlay just filters out ones directed at itself and maps to the keydown.
2. An important piece of this is dispatching an event to the most recently attached overlay. Often, events land on the body with no other context. That's the whole point of tracking overlays as they are attached and detached.

Reply:
Alright, I see. I guess that in most cases the top-most overlay would be the one dispatching the keyboard events, but that won't always be the case (like with `md-autocomplete`). Let's keep it.